src/ore/src/antichain.rs

```diff
@@ -0,0 +1,114 @@
+// Copyright Materialize, Inc. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Antichain utilities.
+
+#![allow(missing_debug_implementations)]
+
+use timely::order::PartialOrder;
+use timely::progress::change_batch::ChangeBatch;
+use timely::progress::frontier::MutableAntichain;
+
+/// A token that holds a mutable antichain and invokes an action should it change.
+pub struct AntichainToken<T: PartialOrder + Ord + Clone> {
+    antichain: MutableAntichain<T>,
+    // This is behind a `Arc` so that .. `Clone` can use `&self` rather than `&mut self`.
+    changes: std::rc::Rc<std::cell::RefCell<ChangeBatch<T>>>,
+    action: std::rc::Rc<std::cell::RefCell<dyn FnMut(std::vec::Drain<(T, i64)>)>>,
+}
+
+impl<T: PartialOrder + Ord + Clone> AntichainToken<T> {
+    /// Creates a new token.
+    pub fn new<I, F>(values: I, action: F) -> Self
+    where
+        I: IntoIterator<Item = T>,
+        F: FnMut(std::vec::Drain<(T, i64)>) + 'static,
+    {
+        let mut token = Self {
+            antichain: MutableAntichain::new(),
+            changes: std::rc::Rc::new(std::cell::RefCell::new(ChangeBatch::new())),
+            action: std::rc::Rc::new(std::cell::RefCell::new(action)),
+        };
+        token.advance(values);
+        token
+    }
+
+    /// Apply direct updates to the maintained antichain.
+    pub fn update<I>(&mut self, updates: I)
+    where
+        I: IntoIterator<Item = (T, i64)>,
+    {
+        self.changes.borrow_mut().extend(updates.into_iter());
+        (self.action.borrow_mut())(
+            self.antichain
+                .update_iter(self.changes.borrow_mut().drain()),
+        );
+    }
+
+    /// Advance the frontier of the mutable antichain to `frontier`.
```

Comment:
If you are interested, this seems like a fine method in which to assert that `self.antichain.less_equal(frontier)`. It would mean that `new` might need to point directly at the `update` method, but it would (I think) rule out a class of errors where someone (accidentally) tries to wind back their antichain. Unless it breaks a bunch of other uses elsewhere, that I'm not noticing at the moment.

Reply:
Done I think? I needed to convert this to a vec so it could be re-used. Is that correct?